// twitch/lib/Toolkit/ObjectTools.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ObjectTools {
    static map(obj, fn) {
        const mapped = Object.entries(obj).map(([key, value]) => {
            let result = {};
            result[key] = fn(value, key);
            return result;
        });
        return Object.assign({}, ...mapped);
    }
    static fromArray(arr, fn) {
        return Object.assign({}, ...arr.map(value => fn(value)));
    }
    // tslint:disable-next-line:no-any
    static indexBy(arr, keyFn) {
        if (typeof keyFn !== 'function') {
            const key = keyFn;
            keyFn = (value) => {
                return value[key].toString();
            };
        }
        return this.fromArray(arr, val => ({ [keyFn(val)]: val }));
    }
    static forEach(obj, fn) {
        Object.entries(obj).forEach(([key, value]) => fn(value, key));
    }
    static entriesToObject(obj) {
        return this.fromArray(obj, ([key, val]) => ({ [key]: val }));
    }
}
exports.default = ObjectTools;
