// twitch/lib/API/TokenInfo.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
const Decorators_1 = require("../Toolkit/Decorators");
class TokenInfo {
    constructor(_data, client) {
        this._data = _data;
        this._client = client;
    }
    get clientId() {
        return this._data.valid ? this._data.client_id : undefined;
    }
    get userId() {
        return this._data.valid ? this._data.user_id : undefined;
    }
    get userName() {
        return this._data.valid ? this._data.user_name : undefined;
    }
    get scopes() {
        return this._data.valid ? this._data.authorization.scopes : [];
    }
    get valid() {
        return this._data.valid;
    }
}
__decorate([
    Decorators_1.NonEnumerable
], TokenInfo.prototype, "_client", void 0);
exports.default = TokenInfo;
