// twitch/lib/API/User/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Decorators_1 = require("../../Toolkit/Decorators");
const ChannelPlaceholder_1 = require("../Channel/ChannelPlaceholder");
const NoSubscriptionProgram_1 = require("../NoSubscriptionProgram");
const NotSubscribed_1 = require("../NotSubscribed");
const NotSubscribed_2 = require("./NotSubscribed");
class User {
    constructor(_data, client) {
        this._data = _data;
        this._client = client;
    }
    get cacheKey() {
        return this._data._id;
    }
    get id() {
        return this._data._id;
    }
    get userName() {
        return this._data.name;
    }
    get displayName() {
        return this._data.display_name;
    }
    get logoUrl() {
        return this._data.logo;
    }
    getChannel() {
        return __awaiter(this, void 0, void 0, function* () {
            return this._client.channels.getChannel(this);
        });
    }
    getChannelPlaceholder() {
        return new ChannelPlaceholder_1.default(this._data._id, this._client);
    }
    getSubscriptionTo(channel) {
        return __awaiter(this, void 0, void 0, function* () {
            return this._client.users.getSubscriptionData(this, channel);
        });
    }
    isSubscribedTo(channel) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.getSubscriptionTo(channel);
                return true;
            }
            catch (e) {
                if (e instanceof NoSubscriptionProgram_1.default || e instanceof NotSubscribed_1.default) {
                    return false;
                }
                throw e;
            }
        });
    }
    getFollows(page, limit, orderBy, orderDirection) {
        return __awaiter(this, void 0, void 0, function* () {
            return this._client.users.getFollowedChannels(this, page, limit, orderBy, orderDirection);
        });
    }
    getFollowTo(channel) {
        return __awaiter(this, void 0, void 0, function* () {
            return this._client.users.getFollowedChannel(this, channel);
        });
    }
    follows(channel) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.getFollowTo(channel);
                return true;
            }
            catch (e) {
                if (e instanceof NotSubscribed_2.default) {
                    return false;
                }
                throw e;
            }
        });
    }
    follow() {
        return __awaiter(this, void 0, void 0, function* () {
            const currentUser = yield this._client.users.getMe();
            return currentUser.followChannel(this);
        });
    }
    unfollow() {
        return __awaiter(this, void 0, void 0, function* () {
            const currentUser = yield this._client.users.getMe();
            return currentUser.unfollowChannel(this);
        });
    }
}
__decorate([
    Decorators_1.NonEnumerable
], User.prototype, "_client", void 0);
exports.default = User;
