// ircv3/lib/Message/MessageTypes/Commands/CapabilityNegotiation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Message_1 = require("../../Message");
class CapabilityNegotiation extends Message_1.default {
    isResponseTo(originalMessage) {
        if (!(originalMessage instanceof CapabilityNegotiation)) {
            return false;
        }
        switch (this.params.command) {
            case 'ACK':
            case 'NAK': {
                // trim is necessary because some networks seem to add trailing spaces (looking at you, Freenode)...
                return originalMessage.params.command === 'REQ'
                    && originalMessage.params.capabilities === this.params.capabilities.trim();
            }
            case 'LS':
            case 'LIST': {
                return originalMessage.params.command === this.params.command;
            }
            default: {
                return false;
            }
        }
    }
    endsResponseTo(originalMessage) {
        if (!(originalMessage instanceof CapabilityNegotiation)) {
            return false;
        }
        switch (this.params.command) {
            case 'LS':
            case 'LIST': {
                return !this.params.continued;
            }
            default: {
                return true;
            }
        }
    }
}
CapabilityNegotiation.COMMAND = 'CAP';
CapabilityNegotiation.PARAM_SPEC = {
    target: {
        match: /^(?:[a-z_\-\[\]\\^{}|`][a-z0-9_\-\[\]\\^{}|`]*|\*)$/i,
        optional: true
    },
    command: {
        match: /^(?:LS|LIST|REQ|ACK|NAK|END|NEW|DEL)$/i,
    },
    version: {
        match: /^\d+$/,
        optional: true
    },
    continued: {
        match: /^\*$/,
        optional: true
    },
    capabilities: {
        trailing: true,
        optional: true
    }
};
CapabilityNegotiation.SUPPORTS_CAPTURE = true;
exports.default = CapabilityNegotiation;
